### ISSUE TITLE
Epic: Block Hooks Features for WP 6.5

### ISSUE THREAD

**Issue by ockham** [WordPress/gutenberg#54904]
- Part of #41236.
- Supersedes #53987.

**Block Hooks have been formerly know as Auto-inserting Blocks.**

## Plans for WordPress 6.5

- [x] [Enable for user-modified templates/parts/patterns](https://core.trac.wordpress.org/ticket/59646)
  - [x] [Store information about known hooked blocks in anchor block metadata](https://core.trac.wordpress.org/ticket/60008)
  - [x] [Block Hooks: Set `ignoredHookedBlocks` metadata upon saving](https://core.trac.wordpress.org/ticket/60506)
- [ ] [No toggle shown for hooked blocks added via filter](https://core.trac.wordpress.org/ticket/59574) ~(Will be be fixed by the enabling Block Hooks for user-modified templates/parts/patterns.)~ _(Edit: [No it won't](https://core.trac.wordpress.org/ticket/59574#comment:4))_
  - [x] https://github.com/WordPress/gutenberg/pull/59396 added it for _modified_ templates/parts/... at least.
- [x] [Block Hooks: Sibling insertion incompatible with constrained layout](https://core.trac.wordpress.org/ticket/60126)
- [x] [Allow child insertion into Navigation block](https://core.trac.wordpress.org/ticket/59743)
- [x] [Allow passing block definitions](https://core.trac.wordpress.org/ticket/59572)
- [x] [Toggle removed if multiple anchor block instances exist, and one hooked block is deleted](https://core.trac.wordpress.org/ticket/59955)
- [ ] #53989
  - Originally touched upon [here](https://github.com/WordPress/gutenberg/issues/39439#issuecomment-1134760286). We might want to give a block some level of control what blocks it allows as its (auto-inserted) children. 
  - There's an `allowedBlock` _attribute_ that's currently used by a number of blocks; AFAICT, it's only handled on the client side, and basically passed through to the `InnerBlocks` component as a prop. _I'm a bit hesitant to use something like this as it  seems currently really up to each individual block, with no standardization, especially on the server (e.g. in the shape of block-supports)._
  - Implementing this later might amount to an API change: Blocks that were previously auto-inserted might cease to be if their parent disallows them 😕 

## Other

- [x] https://github.com/WordPress/wordpress-develop/pull/5399
- [ ] https://github.com/WordPress/gutenberg/pull/58468 (Backport [filter mechanism](https://github.com/WordPress/wordpress-develop/pull/5271) to Gutenberg.)
- [ ] Add higher-level test coverage (integration and/or e2e).
  - Largely follow "Possible scenarios to cover" section in [this comment](https://github.com/WordPress/gutenberg/issues/39439#issuecomment-1521680451).
  - Add small test plugin that includes an auto-inserting block -- probably based on https://github.com/ockham/like-button.
- [ ] Respect template locks.
  - I need to read up on them, but I imagine we shouldn't auto-insert into a locked template? 😬 
- [ ] Respect `"multiple": false`.
  -  Blocks can stipulate they only be inserted once. Auto-insertion should probably respect that; @nerrad made a PoC PR during WCUS Contributor Day: https://github.com/WordPress/gutenberg/pull/53925
  - [Some prior discussion](https://github.com/WordPress/gutenberg/issues/39439#issuecomment-1070683973): "It could also be that autoInsert works upon the first encountered block of the specified type, so it doesn't get repeated if there are multiple navigation blocks in a page. Or maybe this is also a flag like `useOnce`."
  - This still doesn't give control over _which_ navigation block that is (but will default to the first one encountered); we might want a mechanism to select that, too. See next item for a somewhat related problem.
  - Finally, note that the `multiple` field was originally introduced for a post editor context; it isn't really well-defined in a template/Site Editor, as @gziolo [has observed](https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1708031649).
- [ ] Notify user that there’s a new auto-inserted block available (after activating a new plugin that includes that block).
  - In the editor, or even wp-admin? Needs design.
  - How it could be implemented: Make list of registered blocks before and after plugin activation, compare; see if newly added `block.json` has `autoInsert` field? (Doesn't cover blocks that are auto-inserted programmatically.)

**Comment by gziolo** [comment]
One more item to consider would be to potentially iterate over the Plugins section UI in the sidebar panel based on the comments about different expectations from folks that were testing it so far: https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1711906090, https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1712234111 and https://github.com/WordPress/gutenberg/issues/53987#issuecomment-1727927775. I expect it will come back regardless after the call for testing when more people have a chance to play with the integration in the editor.

**Comment by ockham** [comment]
Update:

It was brought to @gziolo's and my attention that the performance of patterns is currently lacking in WP 6.4 (Beta), which is particularly apparent when using TT4. @dmsnell found that `traverse_and_serialize_block(s)` might be a likely culprit, so @gziolo and I looked into potential ways to improve performance. We ended up filing three candidate PRs:

- https://github.com/WordPress/wordpress-develop/pull/5399
- https://github.com/WordPress/wordpress-develop/pull/5413
- https://github.com/WordPress/wordpress-develop/pull/5406

All of them have since been benchmarked and/or profiled, and only one of them turned out to deliver actual improvements.

We'll thus land https://github.com/WordPress/wordpress-develop/pull/5399 in time for Beta 3; there's also https://github.com/WordPress/wordpress-develop/pull/5421 which might improve performance some more (but still needs verifying).

Other than than, @ndiego has done some preliminary testing of Block Hooks; we might try to make some further improvements (or at least file some issues/tickets) based on his results.

**Comment by fabiankaegy** [comment]
Hey all 👋 

I was just trying to create a demo of Block Hooks and it leaves me wondering whether I'm missing something 🤔 

As far as I can tell from the documentation we can _only_ define a block type where the additional elements should get injected. Is this correct? 

I'm asking because I think that may be a pretty large limitation of the API currently. I had though this would work a bit more like CSS selectors where I could "query" for a block insertion point via a more complex selector like `core/post-template core/post-author`

The issue I am running into is that in a block based theme all the available blocks are used for so manny different things that it's almost impossible to make a blanked statement that a new block should be inserted after every author block for example. Because taking the example of the post author block it may be used very differently on a single post template compared to an archve template that renders query loops with the author of each post displayed within. 

This is even more true when thinking about more generalized blocks such as the `core/group` etc. 

Another thing I was hoping for is to target specific variations of a block. For example if one has a custom variation of a product query it would be really cool to say append a block to the post template in the product query variation of the query loop. 

Thanks in advance for any response :) And I get that this is the first MVP version of this. I just want to make sure I'm not missing something :)

**Comment by fabiankaegy** [comment]
Thinking about this more, I think one of the original goals of this was to make it possible for plugins like WooCommerce to automatically inject their "Mini Cart" block into the main navigation of the site. 

However, if I get this right currently, that wouldn't be possible because all you could do is say add it to all navigation menus. But since most themes have way more than one menu that doesn't seem feasible 🤔

**Comment by colorful-tones** [comment]
> As far as I can tell from the documentation we can only define a block type where the additional elements should get injected. Is this correct?

@fabiankaegy I'm not sure I understand this statement, which makes me wonder if I'm missing a larger Block Hooks understanding/requirement for implementation altogether. 

I'm currently trying to test an ACF Block by just dropping in the `blockHooks` and adding in any core block to reference. Is this an unexpected use case?

I'm using `register_block_type()` on `init` and the `block.json` is:

```
{
  "$schema": "https://schemas.wp.org/trunk/block.json",
  "apiVersion": 3,
  "name": "acf/album-meta",
  "version": "0.1.0",
  "title": "ACF Experiment",
  "category": "widgets",
  "icon": "coffee",
  "description": "Just a testing block.",
  "supports": {
    "html": false,
    "mode": false
  },
  "textdomain": "acf-experimental",
  "acf": {
    "renderTemplate": "render.php"
  },
  "usesContext": [
    "postType",
    "postId",
    "commentId"
  ],
  "blockHooks": {
    "core/comment-template": "lastChild"
  }
}
```

I'm running WP 6.4 Beta 3 through WordPress Beta Tester plugin on Local and I have Gutenberg plugin installed and activated. The [Like Button plugin's](https://github.com/ockham/like-button/releases/tag/v0.5.0) block shows fine. However, my block does not display at all.

What am I doing wrong? 🤷

**Comment by fabiankaegy** [comment]
@colorful-tones What I mean by that is that the API currently only supports this: 


```json
{
    "blockHooks": {
        "core/navigation": "lastChild"
    }
}
```

and not

```jsonc
{
    "blockHooks": {
        // core/template-part/header meaning the header variation of the template part block
        "core/template-part/header core/navigation": "lastChild"
    }
}
```

**Comment by colorful-tones** [comment]
@fabiankaegy I see what you mean and that makes sense. I did not expect that level of explicitness for this first iteration of this new API, but I can certainly see how I would want it to get there quickly 😄

**Comment by colorful-tones** [comment]
I've just tried `npx @wordpress/create-block@latest copyright --variant dynamic` and added the following to the `block.json` and this appeared to work fine. So, it is likely something to do with ACF's rendering. I'll try to coordinate with the ACF team on what might be happening. Thanks!

```
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 3,
	"name": "wpe/copyright",
	"version": "0.1.0",
	"title": "Copyright",
	"category": "widgets",
	"icon": "smiley",
	"description": "Example block scaffolded with Create Block tool.",
	"example": {},
	"supports": {
		"html": false
	},
	"textdomain": "copyright",
	"editorScript": "file:./index.js",
	"editorStyle": "file:./index.css",
	"style": "file:./style-index.css",
	"render": "file:./render.php",
	"viewScript": "file:./view.js",
	"blockHooks": {
		"core/comment-template": "lastChild"
	}
}
```

![Screenshot 2023-10-11 at 10 51 07 AM](https://github.com/WordPress/gutenberg/assets/405912/7e6430c1-ff1a-4725-b5ad-c1022b45321f)

**Comment by lgladdy** [comment]
ACF Blocks requires a block's `save` function to be called at least once for the block to work.

In the case of a ACF Block being added via a block hook, unless the template is edited and saved, that never happens - so we abort rendering because we can't figure out the block's type or data. If you make any changes to the template, `save` is fired, so it will work.

This issue is likely specific to ACF Blocks (although, there definitely could be other blocks that setup some data to the block comment on `save` which will also cause issues when hooked).

We'll work around this by inheriting more defaults from `$this`'s `WP_Block` instance passed into the block renderer if they're not set in the block's attributes.

**Comment by ockham** [comment]
Update:

After landing https://github.com/WordPress/wordpress-develop/pull/5399 (which sought to improve Block Hooks performance, see https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1752892056), it was unfortunately discovered that that change [broke template parts](https://github.com/WordPress/gutenberg/issues/55202). We managed to file and land a [fix](https://github.com/WordPress/wordpress-develop/pull/5450) just in time for Beta 3; however, that fix also undid most of the original performance gains, as was found by @felixarntz  😕 

After some discussion with Felix, we then filed a [two-part](https://github.com/WordPress/gutenberg/pull/55217) [fix](https://github.com/WordPress/wordpress-develop/pull/5455) seeking to avoid `theme` attribute injection on the frontend altogether, in order to improve performance at least a bit. We landed both changes on Thursday (the GB fix also necessitated a [package sync](https://github.com/WordPress/wordpress-develop/pull/5468)) in order to have a bit more time to discover potential regressions and/or side effects before RC1.

**Comment by ockham** [comment]
@fabiankaegy @colorful-tones @lgladdy -- Apologies for not engaging in the conversation, I was a bit busy with the [aforementioned issue](https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1763476742) last week. I won't be working tomorrow but hope to catch up with y'all on Tuesday!

**Comment by colorful-tones** [comment]
@ockham I appreciate the update. No apologies necessary. You’re obviously busy and I hope you enjoy your day off. ❤️

**Comment by ockham** [comment]
> Thinking about this more, I think one of the original goals of this was to make it possible for plugins like WooCommerce to automatically inject their "Mini Cart" block into the main navigation of the site.
> 
> However, if I get this right currently, that wouldn't be possible because all you could do is say add it to all navigation menus. But since most themes have way more than one menu that doesn't seem feasible 🤔

We discussed this a bit more [elsewhere](https://github.com/WordPress/gutenberg/issues/54808#issuecomment-1768186965), but since that issue isn't (only) about Block Hooks, I wanted to bring back that conversation here.

FWIW, I succeeded having a Mini Cart block inserted _after_ the Navigation block. The containing Header template part doesn't have any modifications; it seems like the Navigation block falls back to a list of select pages (although not a "complete" page list; possibly a menu created by WooCommerce during setup?)

![mini-cart-after-navigation](https://github.com/WordPress/gutenberg/assets/96308/8eaf9bf1-22b6-4c2c-a381-8dfd52884627)

**Comment by fabiankaegy** [comment]
Thank you for sharing that! :) 

How in the world did the `navigation` block get the `ref` ID added into the block markup whilst the template part is still not customized? :O 

Looking at the actual template part file here: https://github.com/WordPress/wordpress-develop/blob/6528d9840b6800edfb8e75c4924936bb14b3f31b/src/wp-content/themes/twentytwentyfour/parts/header.html#L18

there is no Ref to be found. Since that is env specific :O 

I am really surprised by those results. Really interesting & cool :) Thank you!

**Comment by pagelab** [comment]
Hi, following [this Make post](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/), I did some testing with the PHP `hooked_block_types` filter, using `footer` as `$context->area` but it didn't work in a simple dynamic block plugin.

![2023-10-19 19 29 40](https://github.com/WordPress/gutenberg/assets/1451087/6247dbe6-ffed-4bc0-9466-0141cb0492da)


It's working fine with `header` though.

**Comment by ockham** [comment]
I've added a new section called "Plans for WordPress 6.5" to the issue description (and might update the issue title accordingly), listing TODOs that I discussed with @mtias and @gziolo last week.

**Comment by ockham** [comment]
> How in the world did the `navigation` block get the `ref` ID added into the block markup whilst the template part is still not customized? :O
> 
> Looking at the actual template part file here: https://github.com/WordPress/wordpress-develop/blob/6528d9840b6800edfb8e75c4924936bb14b3f31b/src/wp-content/themes/twentytwentyfour/parts/header.html#L18
> 
> there is no Ref to be found. Since that is env specific :O

Yeah, it is quite curious 🤔 My best guess is that the `ref` is actually inserted on the fly by the block's `edit` component: https://github.com/WordPress/gutenberg/blob/7e3b80c3066c608ff5a00b1337818431f5a3667c/packages/block-library/src/navigation/edit/index.js#L245-L252

The counterpart logic for the frontend seems to be somewhere around here: https://github.com/WordPress/gutenberg/blob/7e3b80c3066c608ff5a00b1337818431f5a3667c/packages/block-library/src/navigation/index.php#L435-L455

---

The Navigation block is a rather tricky beast. FWIW, it's on our TODO list for 6.5 (see updated issue description) to [make it work with `firstChild`/`lastChild` hooked blocks](https://core.trac.wordpress.org/ticket/59743).

**Comment by ockham** [comment]
> Hi, following [this Make post](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/), I did some testing with the PHP `hooked_block_types` filter, using `footer` as `$context->area` but it didn't work in a simple dynamic block plugin.

Thank you for reporting, and apologies for the late reply, @pagelab! I'll try to reproduce this on my end.
In the meantime, and just to eliminate a common source of confusion: Can you confirm that your footer template part doesn't have any user modifications? (See the "Customized" property below.)

![image](https://github.com/WordPress/gutenberg/assets/96308/8780b2e5-6f42-44d3-97ff-aff0de539346)

**Comment by pagelab** [comment]
> Can you confirm that your footer template part doesn't have any user modifications?

@ockham Yes, there were no modifications, as in your picture (Customized: No).

**Comment by ockham** [comment]
Thank you @pagelab! I haven't gotten around to investigating the issue just yet but will do so this week 😄

**Comment by ockham** [comment]
Update:

I got a basic version of hooked blocks insertion into modified templates/parts/patterns to work in https://github.com/WordPress/wordpress-develop/pull/5523. I've started to spin off smaller PRs to land individually; among them https://github.com/WordPress/gutenberg/pull/55811 and https://github.com/WordPress/wordpress-develop/pull/5608, which are ready for review.

I plan to land those PRs this week and will continue to break out PRs, polish the code, and add test coverage.

**Comment by ockham** [comment]
@pagelab I finally managed to test hooked block insertion into a `footer` template part today (together with @SantosGuillamot 👋 ). It seemed to work for us 🤔 

We had one thought: Is it possible that the Group block you're trying to have the dynamic year block inserted into isn't a "direct" child of the Footer template part, but is instead part of a pattern that's used in the Footer? That could explain why it's not working for you, as the filter's `$context` arg is only aware of the "closest" ancestor template, part, or pattern.

Another idea: Since you're trying to insert your block as the `last_child` of a Group block, can you confirm that that Group block already has at least one child block? (The way that child insertion is implemented doesn't work with "empty" parent blocks.)

**Comment by pagelab** [comment]
Sorry for the delay, @ockham; I've missed the notification.

> Is it possible that the Group block you're trying to have the dynamic year block inserted into isn't a "direct" child of the Footer template part but is instead part of a pattern that's used in the Footer?

That is exactly what happened. I'm testing with TT4, and indeed it uses a pattern to define the contents of its default footer template part (`footer.html`).

```html
<!-- wp:pattern {"slug":"twentytwentyfour/footer"} /-->
```

The `header.html` file doesn't do this; it simply adds the code for all blocks directly in the file itself.

> Can you confirm that the Group block already has at least one child block?

Yes, the `patterns/footer.php` file, included in `footer.html`, adds several child blocks.

**Comment by ockham** [comment]
Thank you for confirming, @pagelab (and apologies for _my_ delay 😅 )

Somewhat relieved that it's at least expected behavior; although it indicates that it'd be desirable to maybe provide more context (not just the "immediate" surrounding one) 🤔

**Comment by ockham** [comment]
Update:

A bit less progress in the past couple of weeks as I had to tend to some other things.

I've filed https://github.com/WordPress/wordpress-develop/pull/5712 to supersede https://github.com/WordPress/wordpress-develop/pull/5523, as I [wasn't quite happy](https://github.com/WordPress/wordpress-develop/pull/5609#discussion_r1408393016) with the function signature of `insert_hooked_blocks`.

I'm hoping to land that PR this week, albeit potentially only the parts that persist the information about hooked blocks in the anchor block's `metadata` attribute, but not yet the part that enables hooked blocks insertion for modified layouts. The reason is a UX quirk that @gziolo [pointed out](https://github.com/WordPress/wordpress-develop/pull/5712#issuecomment-1832730396) that that change would introduce, and that we should thus sort out before landing that final part.

I also replied to [some feedback](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/#comment-45867) on the make/core post and filed [one ticket](https://core.trac.wordpress.org/ticket/59955#ticket) based on that feedback.

**Comment by leewillis77** [comment]
I've been looking at making some of our WordPress plugins (particularly WooCommerce extensions) work well in a block-based world, and the Woo devs suggested that dropping some of our feedback here might be helpful. Let me know if you need any more information!

1. I'd echo the sentiment in https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1838425928. For plugins to work reliably and effectively having a full chain of context would definitely be desirable so that we can do things like "insert block X after block Y whenever it appears inside block Z" even when Y isn't necessarily a direct descendent of Z. As well as being more robust, in many cases it would also simplify development.
2. Working out the combination of anchor/context that should be used to insert at a given location is more complex than it feels like it should be as the registration covers *all* registered items, and there's no way (that I could work out) to see what anchor/contexts are triggered on a given page. Not sure what the solution is here, maybe it would be less of an issue if [1] was solved? 
3. Block hooks are only truly useful if they can be used to *reliably* insert blocks into other blocks. At the minute there are some situations where it can be used, and others where it can't. As an example, take the WooCommerce 'product collection' block. Block hooks can be used to register blocks into this *if* the user selects a template, but the hook doesn't take any effect for the default block insertion, which leads to a confusing developer experience, and an inconsistent user experience.

**Comment by ockham** [comment]
Update:

I’ve spent most of last week exploring potential solutions to the [sibling block alignment issue](https://core.trac.wordpress.org/ticket/60126) that was brought to my attention:

I started out with a proof of concept that would set a hooked block’s `layout` attribute — if the block supports it — to the same value of its sibling anchor block (i.e. in case of `before` or `after` insertion) .

Since there’s a risk that that might be too arbitrary or "overbearing", I also explored a correctly. https://github.com/WordPress/wordpress-develop/pull/5835 that a block developer could use to write the code that's needed to set the `layout` attribute.

Finally, I also tried a [variation of the latter approach that would solely extend the existing `hooked_block_types` filter](https://github.com/WordPress/wordpress-develop/pull/5837).

I’ve tried to list pros and cons for the latter [two](https://github.com/WordPress/wordpress-develop/pull/5835#issuecomment-1877355408) [approaches](https://github.com/WordPress/wordpress-develop/pull/5837#issuecomment-1877463590). I’ll continue exploring these options a bit more this week before deciding on how to proceed.

**Comment by ockham** [comment]
Update:

It was a pretty productive week for Block Hooks:
- @tjcafferkey and I finalized [work on the Navigation block](https://github.com/WordPress/gutenberg/pull/57754) and landed it.
- I also got approval for the [Block Hooks toggle behavior](https://github.com/WordPress/gutenberg/pull/57928) change and merged it.
- Finally, the [`hooked_block_{$block_type}` filter PR](https://github.com/WordPress/wordpress-develop/pull/5835) was also approved, so I [committed it to Core](https://core.trac.wordpress.org/changeset/57354).

This is nicely reflected by a number of now-ticked checkboxes in the issue's TODO list 😄 

---

As we're approaching the cut-off date for GB changes to be included in WP 6.5, my focus this week will be on finally unblocking https://github.com/WordPress/wordpress-develop/pull/5726 ([see](https://github.com/WordPress/wordpress-develop/pull/5712#issuecomment-1838758546)).

@swissspidy [suggested adding a non-dynamic version](https://wordpress.slack.com/archives/C02RQBWTW/p1706191528327749?thread_ts=1706190426.390279&cid=C02RQBWTW) of the `hooked_block_{$block_type}` filter (i.e. just `hooked_block`), which I'll also look into.

Finally, @jsnajdr was kind enough to file a PR to introduce an [`allowedBlocks` field in `block.json`](https://github.com/WordPress/gutenberg/pull/58262), which would unblock https://github.com/WordPress/gutenberg/issues/53989, a potential stretch goal for WP 6.5.

**Comment by ockham** [comment]
Update:

The last week saw a bit more of experimentation, with four new PRs opened, only half of which are probably worth pursuing in the short term.

I spent some time last week trying to [update GB's Block Hooks compatibility layer for WP < 6.4](https://github.com/WordPress/gutenberg/pull/58468), which I thought might be a requirement to also add a compat layer for WP < 6.5. While it looks like I got that to work, it also turned out to be a distraction ❌ 

I then also tried to [make hooked blocks insertion into the Navigation block work when used on WP 6.4](https://github.com/WordPress/gutenberg/pull/58508) (rather than `trunk`), as I wanted to highlight this new feature in the release notes for GB 17.6, and it seemed too strong a limitation if people had to run it against WP `trunk`. The approach I chose turned out to be a dead end ❌ 

I then finally started working on [setting the `ignoredHookedBlocks` attribute upon insertion of a new block](https://github.com/WordPress/gutenberg/pull/58553). This is shaping up nicely; one problem I found is while there we have a way of setting `initialAttributes` for blocks inserted from the inserter, that mechanism doesn't apply to a container block's inner blocks, so things get a bit less elegant there 🚧 

While working on this, I discovered that we the client really needs a way to ask the server for a list of hooked blocks for a given anchor block _in a given context_ (e.g. within a Single template, or a header area template part). I spent a bit of time [thinking what a REST API endpoint to that effect would look like](https://core.trac.wordpress.org/ticket/59574#comment:4) and then started [working on one](https://github.com/WordPress/gutenberg/pull/58622) 🚧 

---

GB 17.7 RC 1 will be published on Wednesday, which is the last GB version whose code will be included in WP 6.5 by default. My plan is to finish work on `ignoredHookedBlocks` attribute insertion into new blocks today, and on the `hooked-blocks` endpoint tomorrow (as the latter can be seen as a refinement). This should then allow us to finally unblock hooked blocks insertion into modified templates/parts/patterns.

**Comment by fabiankaegy** [comment]
Now that we are past the Beta 1 cut off point I'm removing this tracking issue from the 6.5 Project board. From now on the board should only contain individual bugs that we want to still fix before 6.5 gets released.

**Comment by ockham** [comment]
Update:

Near the beginning of the past two-week period, @tjcafferkey found and subsequently [fixed](https://github.com/WordPress/gutenberg/pull/58772) a bug that had hooked blocks inserted twice into the Navigation block when running the GB plugin with WP `trunk`.

The focus of the last two weeks was resolving the [mismatch between editor and frontend for newly inserted anchor blocks](https://github.com/WordPress/wordpress-develop/pull/5712#issuecomment-1838758546) in order to unblock injection of hooked blocks into modified templates (https://github.com/WordPress/wordpress-develop/pull/5726).

To that end, @tjcafferkey and I continued work on the client-side [PR to set the `ignoredHookedBlocks` attribute for a newly inserted block](https://github.com/WordPress/gutenberg/pull/58553), and the [`hooked-blocks` controller](https://github.com/WordPress/gutenberg/pull/58622) that would inform the client which blocks were hooked to a given anchor blocks (including those added via a filter).

We managed to merge the client-side PR, which was good enough to unblock hooked blocks injection into modified templates, allowing me to finally [land](https://core.trac.wordpress.org/changeset/57594) https://github.com/WordPress/wordpress-develop/pull/5726 🎉 

However, it became apparent that we wouldn't be able to land the controller in time for WP 6.5's Feature Freeze. We were [considering](https://github.com/WordPress/gutenberg/pull/58622#issuecomment-1936265808) to let Core committers know about the issue and potentially seek "task (blessed)" status for it in order to continue work on it after Beta 1.

That was until last Monday when @gziolo suggested an alternative approach, which turned out much better in a number of ways (didn’t require creating a new controller, kept logic contained on the server side). I managed to get a basic [PR](https://github.com/WordPress/wordpress-develop/pull/6087) up and running by Monday evening, and thanks to Tom’s help, review, and testing, I was able to [land](https://core.trac.wordpress.org/changeset/57627) it just ahead of Beta 1 🙂 

I also [reverted](https://github.com/WordPress/gutenberg/pull/58969) the client-side PR from our previous approach that we’d already merged, since we’re not going to need that now.

---

There's a bit of follow-up work needed now that we've made some changes to how the `ignoredHookedBlocks` attribute is set, chiefly with regard to the Navigation block, for which we already have a [WIP fix](https://github.com/WordPress/gutenberg/pull/59021). There were a few other suggestions by folks that might be too late for 6.5, but I'll file tickets and/or PRs regardless (e.g. allowing returning `null` from `hooked_block_{$block_type}` filter, or adding a generic `hooked_block` version of that filter). Other than that, I'll stand by for bugfixes, and will start work on a Dev Note, time permitting.

**Comment by ockham** [comment]
Update:

We landed the [fix for the Navigation block](https://github.com/WordPress/gutenberg/pull/59021). I then filed tickets and PRs for the feature requests I'd gotten recently:

- https://github.com/WordPress/wordpress-develop/pull/6136 ([ticket](https://github.com/WordPress/wordpress-develop/pull/6136))
- https://github.com/WordPress/wordpress-develop/pull/6143 ([ticket](https://core.trac.wordpress.org/ticket/60580))

Both were fortunately accepted for 6.5 and have already been committed 🎉 

Since https://github.com/WordPress/wordpress-develop/pull/6136 changed the signature of the `hooked_block_{$block_type}` filter, I also released a [new version of the Like Button block plugin](https://github.com/ockham/like-button/releases/tag/v0.8.0) that I use for testing Block Hooks.

Tom then kindly filed a [fix](https://github.com/WordPress/gutenberg/pull/59277) for a [bug](https://github.com/WordPress/gutenberg/issues/59001) that was reported two weeks ago.

---

With WP 6.5 Beta 3 scheduled for tomorrow, I'll prioritize reviewing and landing the [bugfix](https://github.com/WordPress/gutenberg/pull/59277). That aside, I have to start writing the Dev Note for Block Hooks in 6.5. I'll also stand by for other bugs that might crop up.

**Comment by ockham** [comment]
Update:

I filed and merged a [fix](https://github.com/WordPress/gutenberg/pull/59396) to display the Block Hooks toggle based on the presence of the `ignoredHookedBlocks` metadata attribute. I then reviewed @tjcafferkey’s [fix](https://github.com/WordPress/gutenberg/pull/59277) for the Block Hooks toggle when used to insert into the Nav block and [decided](https://github.com/WordPress/gutenberg/pull/59277#pullrequestreview-1900919723) to go with a [simpler alternative](https://github.com/WordPress/gutenberg/pull/59367) (that I had originally advised against) after all. Finally, I filed [another small PR](https://github.com/WordPress/gutenberg/pull/59371) to add a short description to the Plugins (i.e. Block Hooks) panel in the block inspector.

I also published the [Dev Note for Block Hooks in WP 6.5](https://make.wordpress.org/core/2024/03/04/updates-to-block-hooks-in-6-5/).

---

I was alerted to two issues that seem to have been caused by our recent changes to Block Hooks and will need to investigate them: https://github.com/WordPress/gutenberg/issues/59516 and https://core.trac.wordpress.org/ticket/60671.

**Comment by ockham** [comment]
Update:

Our major focus last week was on fixing [two](https://github.com/WordPress/gutenberg/issues/59516) [issues](https://core.trac.wordpress.org/ticket/60671) with garbled entities that were both caused by changes to Block Hooks (one of them affecting the Navigation block, the other one affecting template parts). We managed to fix them both in a similar way, by using a filter instead of an action in order to avoid an additional `wp_update_post` call (which turned out to be the problem): See https://github.com/WordPress/gutenberg/pull/59561 and https://github.com/WordPress/wordpress-develop/pull/6225 for the respective fixes.

---

The fixes are on track for inclusion in RC2, pending a package sync for https://github.com/WordPress/gutenberg/pull/59561, and a backport from Core `trunk` to the `6.5` branch for https://github.com/WordPress/wordpress-develop/pull/6225. I'll take care of the latter today.

There's a minor follow-up work in https://github.com/WordPress/wordpress-develop/pull/6244 that I'll also backport to 6.5. Other than that, I'll keep my eyes open for any other potential bugs. I might also start working/helping with some Block Hooks related documentation.

**Comment by ockham** [comment]
Update:

Last week, I realized that as a side effect of [using the `rest_pre_insert_{$this->post_type}` filter instead of the `rest_after_insert_{$this->post_type}` action](https://github.com/WordPress/wordpress-develop/pull/6225), we were passing the wrong `$context` to Block Hooks filters. I filed a [ticket](https://core.trac.wordpress.org/ticket/60754), and we started working on a [fix](https://github.com/WordPress/wordpress-develop/pull/6254) (which turned out to be a lot of work).

Additionally, it was [reported](https://github.com/WordPress/gutenberg/issues/59867) that creating a new Navigation post object via the REST API didn't work; this was [fixed](https://github.com/WordPress/gutenberg/pull/59875) by @tjcafferkey.

We've started filing a few issues to cover some of the changes we'd like to make to Block Hooks _after_ 6.5, namely:

- [`#60756`](https://core.trac.wordpress.org/ticket/60756): Toggle (re-)inserts hooked block in wrong position (if added by filter)
- [`#60759`](https://core.trac.wordpress.org/ticket/60759): Harmonize ignoredHookedBlocks metadata injection logic
- [`#60769`](https://core.trac.wordpress.org/ticket/60769): Consolidate approach to get the list of hooked blocks

---

With RC3 scheduled for tomorrow, we're hoping to land the fix for the wrong `$context` in time. Time permitting, we might continue to file tickets and issues for Block Hooks past 6.5.

**Comment by gziolo** [comment]
> We've started filing a few issues to cover some of the changes we'd like to make to Block Hooks after 6.5, namely

Do you plan to open a new epic for WP 6.6, or should we put these items in a new section under Block API overview issue?
- #41236

**Comment by annezazu** [comment]
Please create a new epic :D <3  and let's please close this one out when ready!

**Comment by ockham** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/issues/60252.

I've carried over the open issues and tickets from this issue and applied some preliminary gardening. I'll continue to do so over the next couple of days.

---

@leewillis77 Apologies for not replying to [your comment](https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1875449258), I must've somehow missed it. I'll reply shortly over on the new issue!
